designs/2022-languages/README.md

```diff
@@ -0,0 +1,610 @@
+- Repo: eslint/eslint
+- Start Date: 2022-11-25
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: Nicholas C. Zakas
+
+# ESLint Language Plugins
+
+## Summary
+
+This proposal provides a way for plugin authors to define parsers, rules, etc., for any arbitrary programming languages. This would allow ESLint to work on any type of programming language for which there is a plugin defined, not just for JavaScript and its variants.
+
+## Motivation
+
+Over the nine years ESLint has been in existence, it has grown into being not just the most popular JavaScript linter, but arguably the most popular linter for any programming language. Twitter is filled with people lamenting that they don't have an "ESLint for X", where X is some other file type they're working with. Numerous projects have popped up that mimic a lot of what ESLint does at it's core:
+
+* [Stylelint](https://stylelint.io/);
+* [Markdownlint](https://github.com/DavidAnson/markdownlint)
+
+In addition, there are numerous projects that hack around other languages to make them work in ESLint, such as:
+
+* [GraphQL-ESLint](https://github.com/B2o5T/graphql-eslint)
+* [eslint-plugin-json](https://www.npmjs.com/package/eslint-plugin-json)
+* [typescript-eslint](https://typescript-eslint.io)
```

Comment:
If memory serves,  you also needed to create your own scope manager? 

FWIW, `parseForESLint()` will still work, though I think it itself is pretty hacky and would love to move away from it in the future.

Reply:
> If memory serves, you also needed to create your own scope manager?

We did need to, yes, but that's because certain decisions were made early on in the AST definition that caused conflicts with ESLint's rules.

For example - types use `Identifier` nodes for their names. Seems like a good idea in theory, but in practice this is an issue because the scope manager is built fairly naively and just sees most `Identifier`s as possible variable references - causing false positives for `no-unused-vars`, `no-undef`, etc.

We tried augmenting the lint rule with an extra rule (like `eslint-plugin-react/jsx-uses-vars` does) to mark identifiers as used to hide them from `no-unused-vars`, but that was hacky and inaccurate.
A new scope manager was the best option to ensure the ecosystem would work as expected.

FWIW the scope manager is essentially just a fork of `eslint-scope` - adding additional code paths so that the scope analysis can understand type declarations and such (I ended up rewriting most of it from the ground up to modernise the code, but at its core it is a fork as I used `eslint-scope` as a reference).

Reply:
Fun fact - I also did a similar line of work for Flow. [`hermes-eslint`](https://www.npmjs.com/package/hermes-eslint) is a package not unlike typescript-eslint - it enables full support for Flow code in ESLint rules (the hermes parser can parse Flow code and it includes visitor keys, as well as a scope manager that understands flow types)

Reply:
Ah, super interesting, I wasn't aware of `hermes-eslint`.